### ISSUE TITLE
_security API not accepting 2y bcrypt hashes

### ISSUE THREAD

**Issue by ceeeekay** [elastic/elasticsearch#51132]
**Elasticsearch version**: 7.5.1

**Plugins installed**: none

**JVM version**: 13.0.1

**OS version** : Ubuntu 18.04.3 LTS

**Description of the problem including expected versus actual behavior**:

Attempting to use a 2y password_hash to create a user via the `_security` API results in an error

Bcrypt `2a` and `2y` hashes generated by modern tools are functionally identical so I feel that this is a bug. Ubuntu htpasswd, for instance emits `2y` hashes which are not accepted by the API. Attempting to POST a `2y` bcrypt hash results in an error: `hash uses [NOOP]`.

Simply changing the hash prefix to 2a  works, however this means that *generated hashes (e.g., by scripts) can not be posted to the API without first being modified*.

Expected behaviour: The API should accept valid `2y` bcrypt hashes.

**Steps to reproduce**:

1. Create a bcrypt hash with `htpasswd`:
```
$ htpasswd -nBC10 test
New password: <-- test
Re-type new password: 
test:$2y$10$4KUQQvcSYBaQTWUP84ToX.N7CwwgPAT7d0JXofUyH8cH4LuFfT/Y6
```

2. Attempt to create a new user via the _security/user endpoint, using the hash generated in step (1.)

```
POST /_security/user/test
{
  "password_hash" : "$2y$10$4KUQQvcSYBaQTWUP84ToX.N7CwwgPAT7d0JXofUyH8cH4LuFfT/Y6",
  "roles": "some_role"
}

{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Provided password hash uses [NOOP] but the configured hashing algorithm is [BCRYPT]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Provided password hash uses [NOOP] but the configured hashing algorithm is [BCRYPT]"
  },
  "status": 400
}
```
2. Change the hash prefix to `2a`:
```
POST /_security/user/test
{
  "password_hash" : "$2a$10$4KUQQvcSYBaQTWUP84ToX.N7CwwgPAT7d0JXofUyH8cH4LuFfT/Y6",
  "roles": "some_role"
}

{
  "created" : true
}
```

**Comment by jkakavas** [comment]
Thanks for reporting this @ceeeekay . I think you are correct here. `2x` variations indicate specific bcrypt implementation changes in specific libraries/systems and not in the algorithm itself. Thus the output of the hashing should be functionally equivalent for us regardless of what the implementation that created the hash prepends it with. 

We should be able to consume and handle correctly all `2x` variations.

**Comment by tvernum** [comment]
I'm not convinced. For example in PHP `$2x$` hashes are a specific indicator that the hash is incorrectly implemented and may not be compatible with our bcrypt implementation.
I don't think we should blindly accept version identifiers that we don't recognise.

**Comment by jkakavas** [comment]
Reading through a little more, it does look like `2x` versions need the flawed bcrypt version to be able to verify them and `2x`. But `2a`, `2b`, `2y` should be equivalent and should be handled by us. Happy to discuss more in the team meeting

**Comment by tvernum** [comment]
I definitely don't think  the $2y$` issue counts as a bug. The fact that a bunch of tools and libraries decided to start using non standard bcrypt prefixes doesn't make it a bug for us. We don't claim compatibility with hashes generated by other tools, so we're not failing to meet our contract here.

Per my above comment, not accepting `$2x$` is the _right_ behaviour. So I think it _would_ be a bug to accept all `$2?$` hashes and treat them as if they were identical to `$2a$`, and since we can't blindly accept all signfiers, the only way to maintain compatibility with _every possible tool_ would be to track the internet for what eveyone else is doing. We shouldn't set that standard for ourselves.

`$2b$` is arguably a bug (Note: I personally disagree, but I can see a valid argument), since the canonical implementation of bcrypt is OpenBSD, and they switched to the _b_ signifier, so there's a reasonable argument that compatibility with the canonical implementation is a close approximation to standards compliant (given that the bcrypt standard hasn't been updated to reflect real world usage).

In summary: I think what we have is fine, we don't claim to support anything other than ourselves, and we could solve this by making that explicit in the docs. But I'd also be OK with accepting `$2b$`, and if we're going to do that then` $2y$` is fine too. I think accepting any other signifiers is a mistake.

**Comment by CactoT** [comment]
Hi there. I'm taking a look into this. As this is the first time I contribute, couple of questions before wrapping up:

- BCrypt.hashpw still allows the original bcrypt version $2$. Please confirm if I can remove this. Note that the OpenBSD implementation dropped support for it.
- I'm going to move some bcrypt specific code (verifyBcryptHash, for instance) from the Hasher enum to the BCrypt class to improve the separation of responsibilities between them and better align with existing code in BCrypt like hashpw and checkpw. Hasher also has a great deal of code related to PBKDF2, SHA256, etc. Is it ok to also refactor that to move it to a better place or you prefer to leave those things untouched?

Thanks!

**Comment by CactoT** [comment]
@jkakavas @tvernum Could one of you review this, please?

**Comment by tvernum** [comment]
Resolved by https://github.com/elastic/elasticsearch/pull/76083
